### PR TITLE
[WIP] Inject controllerConfigGetter in agent/meterstatus facade

### PR DESCRIPTION
**Pull request by anvial**
This PR injects controllerConfigService in agent/meterstatus facade. Also, it refactors agent/uniter facade to mock controllerConfigGetter function in unit-tests.

To inject this controllerConfigService (controllerConfigGetter), the PR extract `ControllerConfig()` function from `UniStateShim` in common.UnitStateAPI. To do this there was a need to inject `controllerConfigGetter` in several facades where common APIs used:
- `agent/provisioner`
- `agent/upgrader`
- `client/client`
- `client/machinemanager`
- `client/modelmanager`
- `client/modelupdater`

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [x] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
go test github.com/juju/juju/apiserver/facades/agent/meterstatus/... -gocheck.v
go test github.com/juju/juju/apiserver/facades/agent/uniter/... -gocheck.v
go test github.com/juju/juju/apiserver/facades/agent/provisioner/... -gocheck.v 
go test github.com/juju/juju/apiserver/facades/agent/upgrader/... -gocheck.v 
go test github.com/juju/juju/apiserver/facades/client/client/... -gocheck.v 
go test github.com/juju/juju/apiserver/facades/client/machinemanager/... -gocheck.v 
go test github.com/juju/juju/apiserver/facades/client/modelmanager/... -gocheck.v 
go test github.com/juju/juju/apiserver/facades/client/modelupgrader/... -gocheck.v

juju bootstrap lxd lxd
```

### CODE REVIEW

File: apiserver/facades/agent/meterstatus/package_test.go [hunk 1]
```diff
@@ -9,6 +9,8 @@ import (
 	coretesting "github.com/juju/juju/testing"
 )
 
+//go:generate go run go.uber.org/mock/mockgen -package mocks -destination mocks/domain_mock.go github.com/juju/juju/apiserver/facades/agent/meterstatus ControllerConfigGetter
+
```
Comment (SimonRichardson) [inline apiserver/facades/agent/meterstatus/package_test.go:13]: :+1: Thanks for doing this.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 2]
```diff
@@ -134,6 +134,8 @@ func NewProvisionerAPI(ctx facade.Context) (*ProvisionerAPI, error) {
 		return nil, errors.Annotate(err, "instantiating network config API")
 	}
 
+	controllerConfigGetter := ctx.ServiceFactory().ControllerConfig()
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioner.go:137]: We probably want to pull this out at the `register.go` file and just pass the service directly.

File: apiserver/facades/agent/upgrader/upgrader_test.go [hunk 3]
```diff
@@ -34,17 +36,22 @@ type upgraderSuite struct {
...
+	hosted                 *state.State
+	controllerConfigGetter *mocks.MockControllerConfigGetter
 }
 
 var _ = gc.Suite(&upgraderSuite{})
 
 func (s *upgraderSuite) SetUpTest(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	s.controllerConfigGetter = mocks.NewMockControllerConfigGetter(ctrl)
```
Comment (SimonRichardson) [inline apiserver/facades/agent/upgrader/upgrader_test.go:53]: We can't officially do this here as the `gc.C` isn't the same one we pass into the test. This causes much-maligned pain, but the fact is that the error from the test will be incorrect. The gomock controller will be finished at the end of the setup test, not at the end of the test. We need to move this down per test case.

File: apiserver/facades/client/machinemanager/instance_information_test.go [hunk 4]
```diff
@@ -29,12 +29,18 @@ type instanceTypesSuite struct {
...
 }
 
 var _ = gc.Suite(&instanceTypesSuite{})
 
 func (s *instanceTypesSuite) SetUpTest(c *gc.C) {
 	s.authorizer = &apiservertesting.FakeAuthorizer{Tag: names.NewUserTag("admin"), Controller: true}
+
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	s.controllerConfigGetter = mocks.NewMockControllerConfigGetter(ctrl)
```
Comment (SimonRichardson) [inline apiserver/facades/client/machinemanager/instance_information_test.go:43]: Same here, we have to do it, per test. See `setup` in this suite.

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
Nearly there. Just a few issues with testing setup.

File: apiserver/facades/agent/meterstatus/meterstatus_test.go [hunk 5]
```diff
@@ -35,11 +35,16 @@ type meterStatusSuite struct {
...
+
+	controllerConfigGetter *mocks.MockControllerConfigGetter
 }
 
 func (s *meterStatusSuite) SetUpTest(c *gc.C) {
 	s.ApiServerSuite.SetUpTest(c)
 
+	ctrl := gomock.NewController(c)
+	s.controllerConfigGetter = mocks.NewMockControllerConfigGetter(ctrl)
+
```
Comment (SimonRichardson) [inline apiserver/facades/agent/meterstatus/meterstatus_test.go:47]: We can't do this. You have to put these in a `setupMocks` and call it for every method.

I'm starting to think the test suite isn't working for us when we're using mocks.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 6]
```diff
@@ -37,6 +39,23 @@ import (
 	"github.com/juju/juju/storage/poolmanager"
 )
 
+// ControllerConfigGetter defines the methods required to get the controller
+type ControllerConfigGetter interface {
+	ControllerConfig(stdcontext.Context) (controller.Config, error)
+}
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/provisioner.go:45]: I'm thinking these interfaces should be `ControllerConfigService`. So let's try and and standardise on this.

File: apiserver/facades/agent/provisioner/register.go [hunk 7]
```diff
@@ -14,15 +14,24 @@ import (
...
+		return NewProvisionerFacade(ctx) // Relies on agent-set origin in SetHostMachineNetworkConfig.
 	}, reflect.TypeOf((*ProvisionerAPIV11)(nil)))
 }
 
-// newProvisionerAPIV11 creates a new server-side Provisioner API facade.
-func newProvisionerAPIV11(ctx facade.Context) (*ProvisionerAPIV11, error) {
-	provisionerAPI, err := NewProvisionerAPI(ctx)
+// NewProvisionerFacade creates a new server-side Provisioner API facade.
+func NewProvisionerFacade(ctx facade.Context) (*ProvisionerAPI, error) {
+	controllerConfigGetter := ctx.ServiceFactory().ControllerConfig()
```
Comment (SimonRichardson) [inline apiserver/facades/agent/provisioner/register.go:23]: Move this down with the other service factory call.

File: apiserver/facades/schema.json [hunk 8]
```diff
@@ -29619,7 +29619,9 @@
         "Version": 11,
         "AvailableTo": [
             "controller-machine-agent",
-            "machine-agent"
+            "machine-agent",
+            "unit-agent",
+            "model-user"
```
Comment (SimonRichardson) [inline apiserver/facades/schema.json:29624]: Ah, we're clearly changed some order here. There isn't much we can do about this. There is a proposal to fix this that people are happy with, but I don't have the time to do it.

**Review by SimonRichardson — COMMENTED** [review]
It looks ok, so something doesn't like being torn down.

**Comment by anvial** [comment]
Closed in favor https://github.com/juju/juju/pull/16157

**Comment by anvial** [comment]
Closed in favor https://github.com/juju/juju/pull/16157

**Pull request closed without merging**